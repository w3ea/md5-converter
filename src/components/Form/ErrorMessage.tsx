import { FC } from 'react';
import { motion } from 'framer-motion';

type Props = {
    message?: string
}

const ErrorMessage: FC<Props> = ({ message }) => (
    <motion.div
        initial={{ opacity: 0, y: -5 }}
        animate={{ opacity: 1, y: 0, transition: { duration: 0.4 } }}
        exit={{ opacity: 0, y: -5, transition: { duration: 0.4 } }}
        className='absolute bottom-0 left-0 ml-2 mb-3 bg-white flex items-center space-x-1.5'
    >
        {/* Warning icon */}
        <svg className='w-3.5 fill-current text-red-600' xmlns='http://www.w3.org/2000/svg' viewBox='0 0 512 512'>
            <path d='M501.609 384.603L320.543 51.265c-13.666-23.006-37.802-36.746-64.562-36.746-26.76 0-50.896 13.74-64.562 36.746-.103.176-.19.352-.293.528L10.662 384.076c-13.959 23.491-14.223 51.702-.719 75.457 13.535 23.769 37.919 37.948 65.266 37.948h360.544c27.347 0 52.733-14.179 66.267-37.948 13.504-23.754 13.241-51.967-.411-74.93zM225.951 167.148c0-16.586 13.445-30.03 30.03-30.03 16.586 0 30.03 13.445 30.03 30.03v120.121c0 16.584-13.445 30.03-30.03 30.03s-30.03-13.447-30.03-30.03V167.148zm30.03 270.273c-24.839 0-45.046-20.206-45.046-45.046 0-24.839 20.206-45.045 45.046-45.045 24.839 0 45.045 20.206 45.045 45.045.001 24.839-20.205 45.046-45.045 45.046z' />
        </svg>
        <p className='text-3.5 md:text-3.75 text-red-600'>
            {message}
        </p>
    </motion.div>
);

export default ErrorMessage;