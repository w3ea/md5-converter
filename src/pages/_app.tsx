import { FC } from 'react';
import type { AppProps } from 'next/app';
import '../assets/styles/main.scss';
import Main from '../layouts/Main';

const MyApp: FC<AppProps> = ({ Component, pageProps }) => (
    <Main>
        {/* eslint-disable-next-line react/jsx-props-no-spreading */}
        <Component {...pageProps} />
    </Main>
);

export default MyApp;