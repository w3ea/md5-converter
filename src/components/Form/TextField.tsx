import { FC, KeyboardEvent } from 'react';
import { AnimatePresence } from 'framer-motion';
import { UseFormMethods } from 'react-hook-form';
import cls from 'clsx';
import Tooltip from './Tooltip';
import ErrorMessage from './ErrorMessage';

type Props = Partial<Pick<UseFormMethods, 'register' | 'errors'>> & {
    onSubmit: () => Promise<void>
};

const TextField: FC<Props> = ({ register, errors, onSubmit }) => {
    const handleUserKeyPress = (event: KeyboardEvent<HTMLTextAreaElement>) => {
        if (! event.shiftKey && event.key === 'Enter') {
            event.preventDefault();
            return onSubmit();
        }

        return false;
    };

    return (
        <div className='relative flex-grow'>
            <label htmlFor='text' className='text-3.75 text-gray-500 ml-1.5 leading-7'>
                Text
            </label>
            <textarea
                data-testid='text'
                ref={register}
                id='text'
                name='text'
                className={cls('tw-form__text-field', { 'tw-form__text-field--error': errors?.text })}
                onKeyPress={handleUserKeyPress}
                required
            />
            <AnimatePresence>
                {errors?.text && <ErrorMessage message={errors.text.message} />}
            </AnimatePresence>

            <Tooltip />
        </div>
    );
};

export default TextField;