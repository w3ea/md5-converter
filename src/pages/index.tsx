import { FC } from 'react';
import { Provider } from 'react-redux';
import { configureStore } from '@reduxjs/toolkit';
import TableListSlice from '../store/tableListSlice';
import Form from '../components/Form';
import Table from '../components/Table';

export const store = configureStore({ reducer: { tableList: TableListSlice } });

const Home: FC = () => (
    <Provider store={store}>
        <div className='pt-2 md:pt-5 lg:pt-8 pb-20 flex flex-col space-y-18 md:space-y-20'>
            <Form />

            <Table />
        </div>
    </Provider>
);

export default Home;