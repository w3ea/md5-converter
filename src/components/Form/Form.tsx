import { FC, useEffect, useRef, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useForm } from 'react-hook-form';
import { yupResolver } from '@hookform/resolvers/yup';
import * as yup from 'yup';
import { addNewItem, tableListSelector, setTableList } from '../../store/tableListSlice';
import MD5HashGenerator from '../../helpers/MD5HashGenerator';
import TextField from './TextField';
import SubmitButton from './SubmitButton';
import { getFromLocal, saveToLocal } from '../../helpers/localStorage';

type Input = {
    text: string
};

const schema = yup.object().shape({ text: yup.string().required('Text field cannot be empty').trim() });

const Form: FC = () => {
    const [loading, setLoading] = useState(false);
    const dispatch = useDispatch();
    const tableList = useSelector(tableListSelector);
    const didMount = useRef(false);
    const { register, handleSubmit, errors, reset } = useForm<Input>({
        mode: 'onTouched',
        resolver: yupResolver(schema)
    });

    const onSubmit = handleSubmit(async ({ text }) => {
        if (loading) return false;

        setLoading(true);
        const { MD5Hash, error } = await MD5HashGenerator(tableList, text);

        if (error) {
            setLoading(false);
            return false;
        }

        reset();
        dispatch(addNewItem({ originalText: text, MD5Hash: MD5Hash || '' }));
        setLoading(false);
        return false;
    });

    // region get and save tableList to localStorage
    useEffect(() => {
        dispatch(setTableList(getFromLocal()));
    }, [dispatch]);
    useEffect(() => {
        if (didMount.current) {
            saveToLocal(tableList);
        } else {
            didMount.current = true;
        }
    }, [tableList]);
    // endregion

    return (
        <form data-testid='convert-form' className='tw-form' onSubmit={onSubmit}>
            <h1 className='mb-4 md:mb-5 text-5.5 md:text-6 font-bold text-gray-800 text-center'>
                MD5 Converter
            </h1>

            <TextField register={register} errors={errors} onSubmit={onSubmit} />
            <SubmitButton loading={loading} />
        </form>
    );
};

export default Form;