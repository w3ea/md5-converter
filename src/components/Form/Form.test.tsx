import { render, fireEvent, RenderResult } from '@testing-library/react';
import { Provider } from 'react-redux';
import { store } from '../../pages';
import Form from './Form';

window.alert = jest.fn(() => true);

const renderForm = (): RenderResult => render(
    <Provider store={store}>
        <Form />
    </Provider>
);

describe('<Form />', () => {
    it('should match form snapshot', () => {
        const { asFragment } = renderForm();
        expect(asFragment()).toMatchSnapshot();
    });

    it('should display a blank form', () => {
        const { getByTestId } = renderForm();

        expect(getByTestId('convert-form')).toHaveFormValues({ text: '' });
    });

    it('should able to change textarea value', () => {
        const { getByTestId } = renderForm();

        const textarea = getByTestId('text') as HTMLTextAreaElement;

        fireEvent.change(textarea, { target: { value: 'test' } });

        expect(textarea.value).toBe('test');
    });

    it('should display form error on empty field submit', async () => {
        const { getByTestId, findByTestId } = renderForm();

        fireEvent.click(getByTestId('submit-button'));

        expect(await findByTestId('form-error')).toBeInTheDocument();
    });
});