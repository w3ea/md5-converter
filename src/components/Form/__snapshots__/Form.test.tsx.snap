// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Form /> should match form snapshot 1`] = `
<DocumentFragment>
  <form
    class="tw-form"
    data-testid="convert-form"
  >
    <h1
      class="mb-4 md:mb-5 text-5.5 md:text-6 font-bold text-gray-800 text-center"
    >
      MD5 Converter
    </h1>
    <div
      class="relative flex-grow"
    >
      <label
        class="text-3.75 text-gray-600 ml-1.5 leading-7"
        for="text"
      >
        Text
      </label>
      <textarea
        class="tw-form__text-field"
        data-testid="text"
        id="text"
        name="text"
      />
      <div
        class="tooltip"
      >
        <svg
          class="w-5 text-gray-400"
          fill="none"
          stroke="currentColor"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8.228 9c.549-1.165 2.03-2 3.772-2 2.21 0 4 1.343 4 3 0 1.4-1.278 2.575-3.006 2.907-.542.104-.994.54-.994 1.093m0 3h.01M21 12a9 9 0 11-18 0 9 9 0 0118 0z"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          />
        </svg>
        <p
          class="tooltip__text"
        >
          use Shift + Enter for new line
        </p>
      </div>
    </div>
    <div
      class="mt-3 flex md:justify-end"
    >
      <button
        class="tw-form__submit-button"
        data-testid="submit-button"
        type="submit"
      >
        <svg
          class="w-6"
          fill="none"
          stroke="currentColor"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M10.325 4.317c.426-1.756 2.924-1.756 3.35 0a1.724 1.724 0 002.573 1.066c1.543-.94 3.31.826 2.37 2.37a1.724 1.724 0 001.065 2.572c1.756.426 1.756 2.924 0 3.35a1.724 1.724 0 00-1.066 2.573c.94 1.543-.826 3.31-2.37 2.37a1.724 1.724 0 00-2.572 1.065c-.426 1.756-2.924 1.756-3.35 0a1.724 1.724 0 00-2.573-1.066c-1.543.94-3.31-.826-2.37-2.37a1.724 1.724 0 00-1.065-2.572c-1.756-.426-1.756-2.924 0-3.35a1.724 1.724 0 001.066-2.573c-.94-1.543.826-3.31 2.37-2.37.996.608 2.296.07 2.572-1.065z"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          />
          <path
            d="M15 12a3 3 0 11-6 0 3 3 0 016 0z"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          />
        </svg>
        <span
          class="text-4.25"
        >
          Convert
        </span>
      </button>
    </div>
  </form>
</DocumentFragment>
`;
